types/src/metered_channel.rs

```diff
@@ -0,0 +1,295 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+#![allow(dead_code)] // TODO: complete tests - This kinda sorta facades the whole tokio::mpsc::{Sender, Receiver}: without tests, this will be fragile to maintain.
+use futures::{FutureExt, Stream};
+use prometheus::IntGauge;
+use std::task::{Context, Poll};
+use tokio::sync::mpsc::{
+    self,
+    error::{SendError, TryRecvError, TrySendError},
+    OwnedPermit, Permit,
+};
+
+/// An [`mpsc::Sender`](tokio::sync::mpsc::Sender) with an [`IntGauge`]
+/// counting the number of currently queued items.
+#[derive(Debug)]
+pub struct Sender<T> {
+    inner: mpsc::Sender<T>,
+    gauge: IntGauge,
+}
+
+impl<T> Clone for Sender<T> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+            gauge: self.gauge.clone(),
+        }
+    }
+}
+
+/// An [`mpsc::Receiver`](tokio::sync::mpsc::Receiver) with an [`IntGauge`]
+/// counting the number of currently queued items.
+#[derive(Debug)]
+pub struct Receiver<T> {
+    inner: mpsc::Receiver<T>,
+    gauge: IntGauge,
+}
+
+impl<T> Receiver<T> {
+    /// Receives the next value for this receiver.
+    /// Decrements the gauge in case of a successful `recv`.
+    pub async fn recv(&mut self) -> Option<T> {
+        self.inner.recv().inspect(|_| self.gauge.dec()).await
```

Comment:
I believe the `inspect` will get executed even the result is `None` - if that's the case then we shouldn't probably decrement in this case. Can we have a test for this unhappy case?

Reply:
I suggest we simply read the `capacity` and stick it in the gauge rather that trying to carefully march the capacity using `inc` and `dec` which seems fragile to me.

Reply:
Good thought @gdanezis . Reporting capacity might also give us some nice queue semantics (which we could calc in grafana anyways) of "how close am I to deplete?"

Reply:
Except this will only work if all channels have the same capacity, which they really will not have for long. A consequence of having percolating backpressure (my next PR) will be to see the need for different queue sizes.